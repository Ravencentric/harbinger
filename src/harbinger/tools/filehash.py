from hashlib import sha256
from pathlib import Path


def filehash(file: Path, /) -> str:
    """
    Return the sha256 checksum of the given file.
    """
    checksum = sha256()

    buf = bytearray(2**18)
    view = memoryview(buf)

    with file.open("rb") as f:
        while True:
            size = f.readinto(buf)
            if size == 0:
                break  # EOF
            checksum.update(view[:size])

    return checksum.hexdigest().casefold()
