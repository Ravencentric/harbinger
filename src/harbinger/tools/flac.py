from __future__ import annotations

import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from loguru import logger

from ..utils import exe, globber


def flacenc(file: Path, compression: int = 8, destination: Path | None = None, wipe_metadata: bool = True) -> Path:
    """
    Thin flac wrapper that adds support for more codecs via an FFmpeg intermediate.
    """
    if not 0 <= compression <= 8:
        compression = 8

    match destination:
        case None:
            destination = file.with_suffix(".flac")
        case _ if destination.suffix == ".flac":
            destination.parent.mkdir(exist_ok=True, parents=False)
        case _:
            destination.mkdir(exist_ok=True, parents=False)
            destination = destination / f"{file.stem}.flac"

    logger.info(f"Encoding {file.name} to FLAC...")

    try:  # try running natively first
        cmd = [exe("flac"), "-f", f"-{compression}", "-V", file, "-o", destination]
        subprocess.run(cmd, capture_output=True, encoding="utf-8", check=True)  # type: ignore

    except subprocess.CalledProcessError:
        # flac doesn't support the given audio file so we will use FFmpeg to create an intermediate flac
        from tempfile import TemporaryDirectory

        with TemporaryDirectory(ignore_cleanup_errors=True) as tmpdir:
            tempfile = Path(tmpdir) / f"{file.stem}.flac"

            ffmpeg = (
                exe("ffmpeg"),
                "-loglevel",
                "fatal",
                "-i",
                file,
                "-c:a",
                "flac",
                "-compression_level",
                "0",
                tempfile,
            )

            subprocess.run(ffmpeg, capture_output=True, check=True)
            cmd = [exe("flac"), "-f", f"-{compression}", "-V", tempfile, "-o", destination]
            subprocess.run(cmd, capture_output=True, check=True)  # type: ignore

            if wipe_metadata:
                subprocess.run((exe("metaflac"), "--remove-all", destination), capture_output=True, check=True)

    logger.success(f"Successfully encoded {file.name} to FLAC")

    return destination


def concurrent_flacenc(
    src: Path,
    dst: Path | None = None,
    /,
    *,
    compression: int = 8,
    wipe_metadata: bool = True,
    glob: tuple[str, ...] = ("*.wav", "*.w64"),
    recursive: bool = False,
    threads: int | None = None,
) -> tuple[Path, ...]:
    """
    Reference FLAC wrapper with concurrent encoding and support for more codecs via FFmpeg.

    Parameters
    ----------
    src : Path
        Path to the source directory or file.
    dst : Path, optional
        Destination file or directory where transcoded files will be saved.
    compression : int, optional
        Compression level.
    wipe_metadata : bool, optional
        Wipe non-essential metadata.
    glob : tuple[str, ...], optional
        Patterns to match files in the source directory.
    recursive : bool, optional
        Whether to search for files recursively in subdirectories of src.
    threads : int, optional
        Number of threads to use for concurrent encoding.
    """

    if src.is_file():
        return (flacenc(src, compression, dst, wipe_metadata).resolve(),)
    else:
        with ThreadPoolExecutor(max_workers=threads) as executor:
            files = globber(src, glob, recursive)
            futures = [executor.submit(flacenc, file, compression, dst, wipe_metadata) for file in files]
            results = tuple([future.result().resolve() for future in as_completed(futures)])
            executor.shutdown()

        return results
