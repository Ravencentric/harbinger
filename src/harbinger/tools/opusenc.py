from __future__ import annotations

import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from loguru import logger

from ..utils import exe, globber


def get_audio_channels(audio: Path) -> int:
    """
    Utility function to get the channel count from an audio file.
    """
    cmd = [
        exe("ffprobe"),
        audio,
        "-show_entries",
        "stream=channels",
        "-select_streams",
        "a",
        "-of",
        "compact=p=0:nk=1",
        "-v",
        "0",
    ]
    process = subprocess.run(
        cmd,  # type: ignore
        capture_output=True,
        encoding="utf-8",
        check=True,
    )
    return int(process.stdout) if process.stdout.strip().isdigit() else 0


def opusenc(file: Path, bitrate: int | None = None, destination: Path | None = None) -> Path:
    """
    Thin opusenc wrapper that adds automatic bitrate selection and support for more codecs via an FFmpeg intermediate.
    """
    match destination:
        case None:
            destination = file.with_suffix(".opus")
        case _ if destination.suffix in (".opus", ".ogg"):
            destination.parent.mkdir(exist_ok=True, parents=False)
        case _:
            destination.mkdir(exist_ok=True, parents=False)
            destination = destination / f"{file.stem}.opus"

    if bitrate is None:
        match channels := get_audio_channels(file):
            case 1:
                bitrate = 96
            case 2:
                bitrate = 192
            case 7 | 8:
                bitrate = 480
            case _:
                bitrate = 320

    logger.info(f"Encoding {file.name} ({channels} ch) to Opus at {bitrate} kb/s...")

    try:  # try and see if opusenc handles given audio natively
        cmd = [exe("opusenc"), file, "--bitrate", f"{bitrate}", destination]
        subprocess.run(cmd, capture_output=True, encoding="utf-8", check=True)  # type: ignore

    except (
        subprocess.CalledProcessError
    ):  # opusenc doesn't support the given file so we will use FFmpeg to create an intermediate flac
        ffmpeg = (
            exe("ffmpeg"),
            "-loglevel",
            "fatal",
            "-i",
            file,
            "-c:a",
            "flac",
            "-compression_level",
            "0",
            "-f",
            "flac",
            "-",
        )

        pipe = subprocess.Popen(ffmpeg, stdout=subprocess.PIPE)
        cmd = [exe("opusenc"), "-", "--bitrate", f"{bitrate}", destination]
        subprocess.run(cmd, capture_output=True, check=True, stdin=pipe.stdout)  # type: ignore

    logger.success(f"Successfully encoded {file.name} to Opus")

    return destination


def concurrent_opusenc(
    src: Path,
    dst: Path | None = None,
    /,
    *,
    bitrate: int | None = None,
    glob: tuple[str, ...] = ("*.flac", "*.wav", "*.w64"),
    recursive: bool = False,
    threads: int | None = None,
) -> tuple[Path, ...]:
    """
    Opusenc wrapper with concurrent encoding, bitrate selection,
    and support for more codecs via FFmpeg.

    Parameters
    ----------
    src : Path
        Path to the source directory or file.
    dst : Path, optional
        Destination file or directory where transcoded files will be saved.
    bitrate : int, optional
        Target bitrate in kbps.
    glob : tuple[str, ...], optional
        Patterns to match files in the source directory.
    recursive : bool, optional
        Whether to search for files recursively in subdirectories of src.
    threads : int, optional
        Number of threads to use for concurrent encoding.
    """

    if src.is_file():
        return (opusenc(src, bitrate, dst).resolve(),)
    else:
        with ThreadPoolExecutor(max_workers=threads) as executor:
            files = globber(src, glob, recursive)
            futures = [executor.submit(opusenc, file, bitrate, dst) for file in files]
            results = tuple([future.result().resolve() for future in as_completed(futures)])
            executor.shutdown()

        return results
