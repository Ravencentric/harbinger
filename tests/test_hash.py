import subprocess
import sys
from pathlib import Path

from harbinger.tools.filehash import filehash


def test_hash(snapshot):
    assert filehash(Path("tests/__test_data__/Big Buck Bunny - S01E01.ac3")) == snapshot


def test_multiple_hash(snapshot):
    multiple = subprocess.run(
        (
            sys.executable,
            "-m",
            "harbinger",
            "hash",
            "tests/__test_data__/Big Buck Bunny - S01E01.ac3",
            "tests/__test_data__/Big Buck Bunny - S01E01.ac3",
            "--no-table",
            "--no-check",
        ),
        capture_output=True,
        check=True,
    ).stdout.strip()
    assert multiple == snapshot
