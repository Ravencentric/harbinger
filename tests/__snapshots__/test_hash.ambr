# serializer version: 1
# name: test_multiple_hash
  b'Big Buck Bunny - S01E01.ac3: cc42f54bde1b6f2e7e93ab1a74ad0340a1e813f439b48f74d0a287ec83c73e11\r\nBig Buck Bunny - S01E01.ac3: cc42f54bde1b6f2e7e93ab1a74ad0340a1e813f439b48f74d0a287ec83c73e11'
# ---
# name: test_single_hash
  b'cc42f54bde1b6f2e7e93ab1a74ad0340a1e813f439b48f74d0a287ec83c73e11'
# ---
